README.md
# ONAP DCAE policy-handler

See [wiki for DCAE gen2 architecture of policy-handling by DCAE-controller](https://wiki.onap.org/display/DW/DCAE+gen2+architecture+of+policy-handling+by+DCAE-controller)

## web-service for policies to be used by DCAE-Controller

- GET `/policy_latest/<policy_id>` -- get the latest policy from policy-engine that is identified by `policy_id`
- POST `/policies_latest` -- *only for the old(pdp_api_v0) PDP API* -- gets the latest policies that match to the **policy-filter** provided in the body of the request.  The policy-filter mimics the body of the /getConfig on policy-engine.

    sample request - policy-filter

```json
{
  "configAttributes": { "key1":"value1" },
  "configName": "alex_config_name",
  "onapName": "DCAE",
  "policyName": "DCAE_alex.Config_alex_.*",
  "unique": false
}
```

- GET `/healthcheck` - returns 200 OK and current run stats
- **web-socket** to **policy-engine**
  - receives the **push notifications** of the changed and removed policies from the policy-engine,
  - matches the policy-updates to policies and policy-filters found in deployment-handler,
  - retrieves the full policy-bodies of the matched policies,
  - delivers the policy-updates to deployment-handler

## manual http API

- GET `/policies_latest` -- get all the latest policies from policy-engine that either have the policy_id or match to the policy-filter found in deployment-handler deployments
- GET `/catch_up` -- catch up with the latest state of the policy-engine
- GET `/shutdown` -- shutdown the server

----------

## standalone installation

`virtualenv policy_venv`

`cd policy_venv`

`source bin/activate`

`cd ../policy_handler`

`pip install -r requirements.txt`

----------

## preparation to run

`cd policy_venv`

`source bin/activate`

`cd ../policy_handler`

----------

## local configure

local config file `policy_handler/etc/config.json` contains:

```json
{
  "wservice_port" : 25577,
  "consul_url" : "http://consul:8500",
  "consul_timeout_in_secs" : 60,
  "pdp_api_version" : null,
  "policy_handler" : {
    "system" : "policy_handler",
    "tls" : {
      "cert_directory" : "etc/tls/certs/",
      "cacert" : "cacert.pem",
      "private_key" : "key.pem",
      "server_cert" : "cert.pem",
      "server_ca_chain" : "ca_chain.pem"
    }
  },
  "logging" : {...}
}
```

Field descriptions

- `wservice_port` - port of the policy-hanlder web-service
- `consul_url` - optional url for the consul agent
- `consul_timeout_in_secs` - optional timeout in seconds to wait for the response from consul agent
- `pdp_api_version` - optional value for PDP_API_VERSION.
  - The default PDP API is used when this field is null or absent.  The new PDP as of 2019 is the default
  - To use the old PDP API that was created before the end of 2018, put any value like pdp_api_v0 into this field.  Alternatlively, provide a non-empty environment variable $PDP_API_VERSION=pdp_api_v0 on the run of the docker container
- `policy_handler` - local config for policy-handler application
  - `system` - general system name of the policy-handler
  - `tls` - tls settings for the https clients and server - required to enable tls
    - `cert_directory` - relative path pointing to the folder with certificates
    - `cacert` - file name for the ca-cert or ca-bundle file in pem format in cert_directory -- used by https clients
    - `private_key` - file name for the private key in cert_directory -- used by https server
    - `server_cert` - file name for the https server certificate file in pem format in cert_directory
    - `server_ca_chain` - file name for the optional https server ca-chain certificates file in pem format in cert_directory -- used when the ca-chain is not included in the server_cert file
- `logging` - logging config for general logging

----------

## discoverable configure from consul-kv for the **new PDP API**

on the start of the policy-handler, it will get the discoverable part of the configure from consul-kv for the key taken from the `local-config.policy_handler.system`

See [README_pdp_api_v0.md](Readme_pdp_api_v0.md) for instructions on how to set up the policy-handler to work with the **old PDP API** that was created not later than **2018**

### using the new PDP API

As of **R4 Dublin** release, the PDP API is totally redesigned.  The policy-handler is changed to have a startup option to either using the new PDP API, or the old PDP API (pdp_api_v0).

By **default**, the policy-handler will startup configured to use only the **new PDP API**.

#### service_mode in healthcheck

**R4 Dublin**: when the polcy-handler runs against the default **new PDP API**, the /healthchek response should contain the following values under service_mode element

```json
{
  ...
  "service_mode": {
    "is_active_mode_of_operation": false,
    "is_pdp_api_default": true
  }
}

```

#### make sure that the start up of the policy-handler is configured to use the **new PDP API**

make sure that both of the following settings are set properly

- make sure that the environment variable `$PDP_API_VERSION` is either **empty** or **not set** at all on the run of the docker container of the policy-handler

- make sure that the `pdp_api_version` param is either `null` or not present in the startup config of the policy-handler at `etc/config.json`

```json
{
    "pdp_api_version" : null,
}
```

#### the discovarable config of the policy-handler to point to the **new PDP API**

In short: keep the consul-kv record for the policy-handler as before R4 Dublin.

Here is a sample config from consul-kv.  Please replace the {{ ... }} with real setup values

```json
{
  ...
  "policy_engine": {
    "url": "https://{{ policy_ip_addr }}:{{ policy_ip_port }}",
    "path_decision": "/decision/v1",
    "tls_ca_mode": "cert_directory",
    "timeout_in_secs": 60,
    "target_entity": "policy_engine",
    "headers": {
      "Accept": "application/json",
      "Content-Type": "application/json",
      "Authorization": "Basic {{ YOUR_POLICY_ENGINE_AUTHORIZATION }}",
      "ClientAuth": "Basic {{ YOUR_POLICY_ENGINE_CLIENT_AUTH }}",
      "Environment": "{{ YOUR_POLICY_ENGINE_ENVIRONMENT }}"
    }
  }
}
```

----------

## full discoverable configure from consul-kv

```json
{
  "policy_handler": {
    "thread_pool_size": 4,
    "pool_connections": 20,
    "policy_retry_count": 5,
    "policy_retry_sleep": 5,
    "mode_of_operation": "active",
    "catch_up": {
      "interval": 1200
    },
    "reconfigure": {
      "interval": 600
    },
    "policy_engine": {
      "url": "{{ YOUR_POLICY_ENGINE_URL }}",
      "path_decision": "/decision/v1",
      "headers": {
        "Accept": "application/json",
        "Content-Type": "application/json",
        "ClientAuth": "Basic {{ YOUR_POLICY_ENGINE_CLIENT_AUTH }}",
        "Authorization": "Basic {{ YOUR_POLICY_ENGINE_AUTHORIZATION }}",
        "Environment": "{{ YOUR_POLICY_ENGINE_ENVIRONMENT }}"
      },
      "target_entity": "policy_engine",
      "tls_ca_mode": "cert_directory",
      "timeout_in_secs": 60
    },
    "dmaap_mr" : {
        "url" : "http://{{ YOUR_DMAAP_MR_URL }}/events/{{ POLICY_UPDATE_TOPICNAME }}/{{ POLICY_UPDATE_CONSUMEGROUP }}/{{ POLICY_UPDATE_CONSUMERID }}",
        "query": {
          "timeout": 15000
        },
        "headers" : {
            "Content-Type" : "application/json",
            "Authorization": "Basic {{ YOUR_DMAAP_MR_SUBSCRIBER_AUTHORIZATION }}"
        },
        "target_entity" : "dmaap_mr",
        "tls_ca_mode" : "cert_directory",
        "timeout_in_secs": 60
    },
    "deploy_handler": {
      "target_entity": "deployment_handler",
      "url": "http://deployment_handler:8188",
      "max_msg_length_mb": 5,
      "query": {
        "cfy_tenant_name": "default_tenant"
      },
      "tls_ca_mode": "cert_directory",
      "timeout_in_secs": 60
    },
    "service_activator": {
      "target_entity": "service_activator",
      "url": "http://service_activator:123",
      "path_register": "/register",
      "tls_ca_mode": "cert_directory",
      "timeout_in_secs": 20,
      "post_register": {
        "component_name": "policy_handler",
        "reconfigure_path": "/reconfigure",
        "http_protocol": "http"
      }
    }
  }
}
```

### field description in yaml format that is equivalent to the actual json structure of the full discoverable config

```yaml
  policy_handler :
    # parallelize the getConfig queries to policy-engine on each policy-update notification
    thread_pool_size : 4

    # parallelize requests to policy-engine and keep them alive
    pool_connections : 20

    # retry to getConfig from policy-engine on policy-update notification
    policy_retry_count : 5
    policy_retry_sleep : 5

    # mode of operation for the policy-handler
    # either active or passive
    # in passive mode the policy-hanlder will not listen to
    #                 and will not bring the policy-updates from policy-engine
    mode_of_operation : "active"

    # config of automatic catch_up for resiliency
    catch_up :
      # interval in seconds on how often to call automatic catch_up
      # example: 1200 is 20*60 seconds that is 20 minutes
      interval : 1200

    # config of periodic reconfigure-rediscover for adaptability
    reconfigure:
      # interval in seconds on how often to call automatic reconfigure
      # example: 600 is 10*60 seconds that is 10 minutes
      interval : 600

    # PDP (policy-engine) config
    # These are the url of and the auth for the external system, namely the policy-engine (PDP).
    # We obtain that info manually from PDP folks at the moment.
    # In long run we should figure out a way of bringing that info into consul record
    #    related to policy-engine itself.
    policy_engine :
      url : "{{ YOUR_POLICY_ENGINE_URL }}"
      # path to decision on the new PDP API as of 2019
      path_decision : "/decision/v1"
      headers :
        Accept : "application/json"
        "Content-Type" : "application/json"
        ClientAuth : "Basic {{ YOUR_POLICY_ENGINE_CLIENT_AUTH }}"

        # to override the Authorization value,
        #    set the environment vars $PDP_USER and $PDP_PWD in policy-handler
        Authorization : "Basic {{ YOUR_POLICY_ENGINE_AUTHORIZATION }}"

        Environment : "{{ YOUR_POLICY_ENGINE_ENVIRONMENT }}"

      # target_entity name that is used for logging
      target_entity : "policy_engine"

      # optional tls_ca_mode specifies where to find the cacert.pem for tls
      #   can be one of these:
      #       "cert_directory" - use the cacert.pem stored locally in cert_directory.
      #                          this is the default if cacert.pem file is found
      #
      #       "os_ca_bundle"     - use the public ca_bundle provided by linux system.
      #                          this is the default if cacert.pem file not found
      #
      #       "do_not_verify"  - special hack to turn off the verification by cacert and hostname
      tls_ca_mode : "cert_directory"
      # optional timeout_in_secs specifies the timeout for the http requests
      timeout_in_secs: 60


    # DMaaP MR subscriber config
    # These are the url of and the auth for the external system, namely the policy-engine (PDP).
    # We obtain that info manually from PDP and DMaaP folks at the moment.
    dmaap_mr :
      url: "http://{{ YOUR_DMAAP_MR_URL }}/events/{{ POLICY_UPDATE_TOPICNAME }}/{{ POLICY_UPDATE_CONSUMEGROUP }}/{{ POLICY_UPDATE_CONSUMERID }}"

      query:
        # The number of milliseconds for DMaaP MR to wait for messages if none are immediately available.
        # This should normally be used, and set at 15000 or higher.
        # This is referred to as long-polling timeout
        # ?timeout=15000 passed to DMaaP MR in the query
        timeout: 15000

      headers:
        "Content-Type": "application/json"
        # provide Authorization for the subscriber if using https and user-password authentication
        # to override the Authorization value,
        #    set the environment vars $DMAAP_MR_USER and $DMAAP_MR_PWD in policy-handler
        Authorization: "Basic {{ YOUR_DMAAP_MR_SUBSCRIBER_AUTHORIZATION }}"

      # target_entity name that is used for logging
      target_entity: "dmaap_mr"
      # optional tls_ca_mode specifies where to find the cacert.pem for tls
      #   can be one of these:
      #       "cert_directory" - use the cacert.pem stored locally in cert_directory.
      #                          this is the default if cacert.pem file is found
      #
      #       "os_ca_bundle"     - use the public ca_bundle provided by linux system.
      #                          this is the default if cacert.pem file not found
      #
      #       "do_not_verify"  - special hack to turn off the verification by cacert and hostname
      tls_ca_mode: "cert_directory"
      # optional timeout_in_secs specifies the timeout for the http requests
      timeout_in_secs: 60


    # deploy_handler config
    #    changed from string "deployment_handler" in 2.3.1 to structure in 2.4.0
    deploy_handler :
      # name of deployment-handler service used by policy-handler for logging
      target_entity : "deployment_handler"
      # url of the deployment-handler service for policy-handler to direct the policy-updates to
      #   - expecting dns to resolve the name deployment_handler to ip address
      url : "http://deployment_handler:8188"
      # limit the size of a single data segment for policy-update messages
      #       from policy-handler to deployment-handler in megabytes
      max_msg_length_mb : 5
      query :
        # optionally specify the tenant name for the cloudify under deployment-handler
        #    if not specified the "default_tenant" is used by the deployment-handler
        cfy_tenant_name : "default_tenant"
      # optional tls_ca_mode specifies where to find the cacert.pem or skip tls verification
      #   can be one of these:
      #       "cert_directory" - use the cacert.pem stored locally in cert_directory.
      #                          this is the default if cacert.pem file is found
      #
      #       "os_ca_bundle"     - use the public ca_bundle provided by linux system.
      #                          this is the default if cacert.pem file not found
      #
      #       "do_not_verify"  - special hack to turn off the verification by cacert and hostname
      tls_ca_mode : "cert_directory"
      # optional timeout_in_secs specifies the timeout for the http requests
      timeout_in_secs: 60

    # optional service_activator config
    #    is used to report the active-passive mode_of_operation of the DCAE-C cluster
    service_activator :
      # name of service_activator service used by policy-handler for logging
      target_entity : "service_activator"
      # url of the service_activator service for policy-handler to detect the mode-of-operation
      url : "http://service_activator:123"
      # path-endpoint to posting the registration to get the mode_of_operation
      path_register : "/register"
      # optional tls_ca_mode specifies where to find the cacert.pem or skip tls verification
      #   can be one of these:
      #       "cert_directory" - use the cacert.pem stored locally in cert_directory.
      #                          this is the default if cacert.pem file is found
      #
      #       "os_ca_bundle"     - use the public ca_bundle provided by linux system.
      #                          this is the default if cacert.pem file not found
      #
      #       "do_not_verify"  - special hack to turn off the verification by cacert and hostname
      tls_ca_mode : "cert_directory"
      # optional timeout_in_secs specifies the timeout for the http requests
      timeout_in_secs : 20
      # /register request message to post to the service_activator
      # put anything that service_activator expects for the registration of the policy-handler
      post_register :
        # discoverable component name
        component_name : "policy_handler"
        # endpoint on policy-handler that will receive the POST on reconfigure event
        reconfigure_path : "/reconfigure"
        # protocol for the /reconfigure event
        http_protocol : "http"
```

----------

## run

in folder `policy_handler`:

`./run_policy.sh`

----------

## customization per company

### ```etc_customize/``` folder

- company is expected to place any company specific files required to be in the docker image in the folder ```etc_customize/```

- change the ```etc_customize/customize.sh``` script to perform company specific actions during docker image build

- ```etc_customize/customize.sh``` script is expected to be overridden by company to customize docker image build

### ```policyhandler/customize/``` folder

contains ```CustomizerBase``` and ```Customizer``` classes

- ```CustomizerBase``` defines the interface and the default=ONAP behavior

- ```CustomizerBase``` is owned by ONAP and should not be changed by the company

- ```Customizer``` inherits ```CustomizerBase```

- policy-handler instantiates ```Customizer``` to get the customized behavior

- ```Customizer``` is owned by the company and should be changed by the company
- ONAP is not going to change ```Customizer```

- the methods of ```Customizer``` are expected to be overridden by the company to change the behavior of the policy-handler

- samples are provided for methods in ```Customizer``` class as the commented out lines

- Company is allowed to add more files to customize/ folder if that is required for better structuring of their code as soon as it is invoked by the methods of ```Customizer```

here is an example of ```customizer.py```

```python
"""contains the Customizer class with method overrides per company specification"""

from .customizer_base import CustomizerBase

class Customizer(CustomizerBase):
    """
    the Customizer class inherits CustomizerBase that is owned by ONAP

    :Customizer: class is owned by the company that needs to customize the policy-handler

    :override: any method defined in the CustomizerBase class to customize the behavior of the policy-handler
    """
    def __init__(self):
        """class that contains the customization"""
        super().__init__()

    def get_service_url(self, audit, service_name, service):
        """
        returns the service url when called from DiscoveryClient

        this is just a sample code - replace it with the real customization
        """
        service_url = super().get_service_url(audit, service_name, service)
        audit.info("TODO: customization for service_url on {0}".format(service_name))
        return service_url
```

----------


Changelog.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [5.1.3] - 2021-07-01
### Changed
         - [DCAEGEN2-2849](https://jira.onap.org/browse/DCAEGEN2-2849) - Remove VOLUME definition in Dockerfile to prevent startup failure in recent versions of containerd.  (See also https://github.com/containerd/containerd/issues/5547)
## [5.1.2] - 03/03/2021
### Changed
         - [DCAEGEN2-2652](https://jira.onap.org/browse/DCAEGEN2-2652) - Policy Handler startup error; CherryPy/requirements update

## [5.1.1] - 03/02/2021


policyhandler/config.py
# ================================================================================
# Copyright (c) 2017-2020 AT&T Intellectual Property. All rights reserved.
# ================================================================================
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============LICENSE_END=========================================================
#

"""read and use the config"""

import base64
import copy
import json
import logging
import logging.config
import os

from .onap.audit import Audit
from .utils import Utils

LOGS_DIR = 'logs'

try:
    os.makedirs(LOGS_DIR, mode=0o770, exist_ok=True)
except Exception:
    pass

logging.basicConfig(
    filename=os.path.join(LOGS_DIR, 'policy_handler.log'),
    format=('%(asctime)s.%(msecs)03d %(levelname)+8s ' +
            '%(threadName)s %(name)s.%(funcName)s: %(message)s'),
    datefmt='%Y%m%d_%H%M%S', level=logging.DEBUG)

_LOGGER = Utils.get_logger(__file__)

class Settings(object):
    """settings of module or an application
    that is the config filtered by the collection of config-keys.

    keeps track of changes versus the previous set_config unless committed
    """
    def __init__(self, *config_keys):
        """provide the collection of top level keys in config to limit the config"""
        self._config_keys = config_keys
        self._changed = False
        self._config = None
        self._prev_config = None

    def __str__(self):
        """get str of the config"""
        if not self._changed:
            return Audit.json_dumps({
                "config_keys": self._config_keys,
                "config": self._config
            })

        return Audit.json_dumps({
            "config_keys": self._config_keys,
            "changed": self._changed,
            "config": self._config,
            "prev_config": self._prev_config
        })

    def is_loaded(self):
        """whether loaded already"""
        return bool(self._config)

    def commit_change(self):
        """set the prev config to the latest config"""
        self._prev_config = copy.deepcopy(self._config)
        self._changed = False

    def _set_changed(self):
        """determine whether the config changed"""
        self._changed = not (self._prev_config
                             and Utils.are_the_same(self._prev_config, self._config,
                                                    Audit.json_dumps))

    def set_config(self, config, auto_commit=False):
        """update the config"""
        self.commit_change()

        if isinstance(config, Settings):
            config = config._config

        if not isinstance(config, dict):
            config = {}

        self._config = copy.deepcopy(dict((k, v) for (k, v) in config.items()
                                          if not self._config_keys or k in self._config_keys))

        if auto_commit:
            self.commit_change()
        else:
            self._set_changed()

    def is_changed(self):
        """whether the config has changed"""
        return self._changed

    def get_by_key(self, config_key, default=None):
        """get the latest sub config by config_key and whether it has changed"""
        if not config_key or not isinstance(config_key, str):
            return False, default

        value = copy.deepcopy(self._config.get(config_key, default))
        if not self._prev_config:
            return True, value
        prev_value = self._prev_config.get(config_key, default)
        return self._changed and not Utils.are_the_same(prev_value, value, Audit.json_dumps), value

    def update(self, config_key, value=None):
        """set the latest sub config by config_key and determine whether the config has changed"""
        if not config_key:
            return

        self._config[config_key] = copy.deepcopy(value)
        self._set_changed()


class Config(object):
    """main config of the application"""
    CONFIG_FILE_PATH = "etc/config.json"
    LOGGER_CONFIG_FILE_PATH = "etc/common_logger.config"
    SERVICE_NAME_POLICY_HANDLER = "policy_handler"

    FIELD_SYSTEM = "system"
    FIELD_CONSUL_URL = "consul_url"
    FIELD_WSERVICE_PORT = "wservice_port"
    FIELD_TLS = "tls"
    FIELD_POLICY_ENGINE = "policy_engine"
    DMAAP_MR = "dmaap_mr"
    POOL_CONNECTIONS = "pool_connections"
    DEPLOY_HANDLER = "deploy_handler"
    THREAD_POOL_SIZE = "thread_pool_size"
    POLICY_RETRY_COUNT = "policy_retry_count"
    POLICY_RETRY_SLEEP = "policy_retry_sleep"
    RECONFIGURE = "reconfigure"
    TIMER_INTERVAL = "interval"
    REQUESTS_VERIFY = "verify"
    TLS_CA_MODE = "tls_ca_mode"
    TLS_WSS_CA_MODE = "tls_wss_ca_mode"
    TLS_CA_MODE_DO_NOT_VERIFY = "do_not_verify"
    TIMEOUT_IN_SECS = "timeout_in_secs"
    CONSUL_TIMEOUT_IN_SECS = "consul_timeout_in_secs"
    WS_PING_INTERVAL_IN_SECS = "ws_ping_interval_in_secs"
    DEFAULT_TIMEOUT_IN_SECS = 60
    SERVICE_ACTIVATOR = "service_activator"
    MODE_OF_OPERATION = "mode_of_operation"
    PDP_API_VERSION = "PDP_API_VERSION"
    QUERY_TIMEOUT = "timeout"
    PDP_USER = "PDP_USER"
    PDP_PWD = "PDP_PWD"
    DMAAP_MR_USER = "DMAAP_MR_USER"
    DMAAP_MR_PWD = "DMAAP_MR_PWD"

    system_name = SERVICE_NAME_POLICY_HANDLER
    wservice_port = 25577
    _pdp_api_version = os.environ.get(PDP_API_VERSION)
    consul_url = "http://consul:8500"
    consul_timeout_in_secs = DEFAULT_TIMEOUT_IN_SECS
    tls_cacert_file = None
    tls_server_cert_file = None
    tls_private_key_file = None
    tls_server_ca_chain_file = None
    _pdp_authorization = None
    _dmaap_mr_authorization = None

    _local_config = Settings()
    discovered_config = Settings()

    @staticmethod
    def _get_tls_file_path(tls_config, cert_directory, tls_name):
        """calc file path and verify its existance"""
        file_name = tls_config.get(tls_name)
        if not file_name:
            return None
        tls_file_path = os.path.join(cert_directory, file_name)
        if not os.path.isfile(tls_file_path) or not os.access(tls_file_path, os.R_OK):
            _LOGGER.error("invalid %s: %s", tls_name, tls_file_path)
            return None
        return tls_file_path

    @staticmethod
    def _set_tls_config(tls_config):
        """verify and set tls certs in config"""
        try:
            Config.tls_cacert_file = None
            Config.tls_server_cert_file = None
            Config.tls_private_key_file = None
            Config.tls_server_ca_chain_file = None

            if not (tls_config and isinstance(tls_config, dict)):
                _LOGGER.info("no tls in config: %s", json.dumps(tls_config))
                return

            cert_directory = tls_config.get("cert_directory")

            if not (cert_directory and isinstance(cert_directory, str)):
                _LOGGER.warning("unexpected tls.cert_directory: %r", cert_directory)
                return

            cert_directory = os.path.join(
                os.path.dirname(os.path.dirname(os.path.realpath(__file__))), cert_directory)
            if not (cert_directory and os.path.isdir(cert_directory)):
                _LOGGER.warning("ignoring invalid cert_directory: %s", cert_directory)
                return

            Config.tls_cacert_file = Config._get_tls_file_path(tls_config, cert_directory, "cacert")
            Config.tls_server_cert_file = Config._get_tls_file_path(tls_config, cert_directory,
                                                                    "server_cert")
            Config.tls_private_key_file = Config._get_tls_file_path(tls_config, cert_directory,
                                                                    "private_key")
            Config.tls_server_ca_chain_file = Config._get_tls_file_path(tls_config, cert_directory,
                                                                        "server_ca_chain")

        finally:
            _LOGGER.info("tls_cacert_file = %s", Config.tls_cacert_file)
            _LOGGER.info("tls_server_cert_file = %s", Config.tls_server_cert_file)
            _LOGGER.info("tls_private_key_file = %s", Config.tls_private_key_file)
            _LOGGER.info("tls_server_ca_chain_file = %s", Config.tls_server_ca_chain_file)

    @staticmethod
    def init_config(file_path=None):
        """read and store the config from config file"""
        if Config._local_config.is_loaded():
            _LOGGER.info("config already inited: %s", Config._local_config)
            return

        if not file_path:
            file_path = Config.CONFIG_FILE_PATH

        loaded_config = None
        if os.access(file_path, os.R_OK):
            with open(file_path, 'r') as config_json:
                loaded_config = json.load(config_json)

        if not loaded_config:
            _LOGGER.warning("config not loaded from file: %s", file_path)
            return

        _LOGGER.info("config loaded from file(%s): %s", file_path, Audit.json_dumps(loaded_config))
        logging_config = loaded_config.get("logging")
        if logging_config:
            logging.config.dictConfig(logging_config)

        Config.wservice_port = loaded_config.get(Config.FIELD_WSERVICE_PORT, Config.wservice_port)
        Config.consul_url = os.environ.get(
            "CONSUL_URL", loaded_config.get(Config.FIELD_CONSUL_URL, Config.consul_url)).rstrip("/")
        Config.consul_timeout_in_secs = loaded_config.get(Config.CONSUL_TIMEOUT_IN_SECS)
        if not Config.consul_timeout_in_secs or Config.consul_timeout_in_secs < 1:
            Config.consul_timeout_in_secs = Config.DEFAULT_TIMEOUT_IN_SECS

        Config._pdp_api_version = os.environ.get(
            Config.PDP_API_VERSION, loaded_config.get(Config.PDP_API_VERSION.lower()))

        pdp_user = os.environ.get(Config.PDP_USER)
        pdp_pwd = os.environ.get(Config.PDP_PWD)
        if pdp_user and pdp_pwd:
            Config._pdp_authorization = "Basic {}".format(base64.b64encode(
                ("{}:{}".format(pdp_user, pdp_pwd)).encode()).decode("utf-8"))

        dmaap_mr_user = os.environ.get(Config.DMAAP_MR_USER)
        dmaap_mr_pwd = os.environ.get(Config.DMAAP_MR_PWD)
        if dmaap_mr_user and dmaap_mr_pwd:
            Config._dmaap_mr_authorization = "Basic {}".format(base64.b64encode(
                ("{}:{}".format(dmaap_mr_user, dmaap_mr_pwd)).encode()).decode("utf-8"))

        local_config = loaded_config.get(Config.SERVICE_NAME_POLICY_HANDLER, {})
        Config.system_name = local_config.get(Config.FIELD_SYSTEM, Config.system_name)

        Config._set_tls_config(local_config.get(Config.FIELD_TLS))

        Config._local_config.set_config(local_config, auto_commit=True)

    @staticmethod
    def _overwrite_discovered_config(audit, discovered_config):
        """replace the secrets in discovered_config with data from environment"""
        changes = []
        if Config._pdp_authorization:
            pdp_cfg = discovered_config.get("policy_engine", {})
            if pdp_cfg.get("url", "").lower().startswith("https:"):
                pdp_cfg.get("headers", {})["Authorization"] = Config._pdp_authorization
                changes.append("pdp_authorization")

        if Config._dmaap_mr_authorization:
            dmaap_mr_cfg = discovered_config.get("dmaap_mr", {})
            if dmaap_mr_cfg.get("url", "").lower().startswith("https:"):
                dmaap_mr_cfg.get("headers", {})["Authorization"] = Config._dmaap_mr_authorization
                changes.append("dmaap_mr_authorization")

        if changes:
            _LOGGER.info(audit.info("overwritten discovered config: {}".format(", ".join(changes))))


    @staticmethod
    def discover(audit):
        """bring the config settings from the discovery service"""
        discovery_key = Config.system_name
        from .discovery import DiscoveryClient
        new_config = DiscoveryClient.get_value(audit, discovery_key)

        if not new_config or not isinstance(new_config, dict):
            _LOGGER.warning(audit.warn("unexpected config from discovery: {}".format(new_config)))
            return

        _LOGGER.debug(audit.debug("loaded config from discovery({}): {}".format(
            discovery_key, Audit.json_dumps(new_config))))
        discovered_config = new_config.get(Config.SERVICE_NAME_POLICY_HANDLER)

        Config._overwrite_discovered_config(audit, discovered_config)

        Config.discovered_config.set_config(discovered_config)
        _LOGGER.info(audit.info("config from discovery: {}".format(Config.discovered_config)))


    @staticmethod
    def get_tls_verify(tls_ca_mode=None):
        """
        generate verify value based on tls_ca_mode

        tls_ca_mode can be one of:

        "cert_directory" - use the cacert.pem stored locally in cert_directory.
        this is the default if cacert.pem file is found

        "os_ca_bundle" - use the public ca_bundle provided by linux system.
        this is the default if cacert.pem file not found

        "do_not_verify" - special hack to turn off the verification by cacert and hostname
        """
        if tls_ca_mode == Config.TLS_CA_MODE_DO_NOT_VERIFY:
            return False

        if tls_ca_mode == "os_ca_bundle" or not Config.tls_cacert_file:
            return True

        return Config.tls_cacert_file

    @staticmethod
    def get_requests_kwargs(tls_ca_mode=None):
        """generate kwargs with verify for requests based on the tls_ca_mode"""
        return {Config.REQUESTS_VERIFY: Config.get_tls_verify(tls_ca_mode)}

    @staticmethod
    def is_pdp_api_default(log_status=True):
        """whether to use the old (2018) or the default pdp API (started in 2019)"""
        is_default = (Config._pdp_api_version is None)
        if log_status:
            _LOGGER.info("_pdp_api_version(%s) default(%s)", Config._pdp_api_version, is_default)
        return is_default
